deps/rabbit/src/rabbit_peer_discovery.erl

```diff
@@ -1001,19 +1015,35 @@ maybe_unregister() ->
             ok
     end.
 
--spec discovery_retries() -> {Retries, RetryDelay} when
-      Retries :: non_neg_integer(),
+-spec discovery_retries(Backend) -> {Retries, RetryDelay} when
+      Backend :: backend(),
+      Retries :: non_neg_integer() | infinity,
       RetryDelay :: non_neg_integer().
 
-discovery_retries() ->
-    case application:get_env(rabbit, cluster_formation) of
-        {ok, Proplist} ->
-            Retries  = proplists:get_value(discovery_retry_limit,    Proplist, ?DEFAULT_DISCOVERY_RETRY_COUNT),
-            Interval = proplists:get_value(discovery_retry_interval, Proplist, ?DEFAULT_DISCOVERY_RETRY_INTERVAL_MS),
-            {Retries, Interval};
-        undefined ->
-            {?DEFAULT_DISCOVERY_RETRY_COUNT, ?DEFAULT_DISCOVERY_RETRY_INTERVAL_MS}
-    end.
+discovery_retries(Backend) ->
+    _ = code:ensure_loaded(Backend),
+    {Retries0, Interval} = case application:get_env(rabbit, cluster_formation) of
+                               {ok, Proplist} ->
+                                   Retries1  = proplists:get_value(
+                                                 discovery_retry_limit,
+                                                 Proplist,
+                                                 ?DEFAULT_DISCOVERY_RETRY_COUNT),
+                                   Interval1 = proplists:get_value(
+                                                 discovery_retry_interval,
+                                                 Proplist,
+                                                 ?DEFAULT_DISCOVERY_RETRY_INTERVAL_MS),
+                                   {Retries1, Interval1};
+                               undefined ->
+                                   {?DEFAULT_DISCOVERY_RETRY_COUNT, ?DEFAULT_DISCOVERY_RETRY_INTERVAL_MS}
+                           end,
+    Retries = case erlang:function_exported(Backend, retry_forever, 0)
+                   andalso Backend:retry_forever() of
+                  true ->
+                      infinity;
+                  false ->
+                      Retries0
+              end,
+    {Retries, Interval}.
```

Comment:
I find the logic to be difficult to reason about. The retry is configurable except when it’s not if the backend says to retry forever. Without looking thoroughly at the code, I would expect the configuration to take precedence over the backend wanted behavior. That’s just because the parameter always did that and the function starts with that parameter.

What about changing the semantic slightly:
1. Have a `retry_strategy() -> limited | unlimited` (the lack of this callback would mean `limited`).
2. Read the configuration parameter if `limited`, uses default values if not configured.

The result is the same, but my point is that the function should make it clear that the backend decides first if the retry is configurable, then we possibly read the parameter.

Reply:
Also if the backend tell it’s not configurable but something is configured, perhaps we should log an info.

Reply:
As we already said, it should be cleaner if you just call the callback and catch the `undef` exception.

Reply:
1. I've implemented the logic you suggested (hopefully :) )
2. I've added config validation in `rabbit_peer_discovery_k8s:init/0`
3. I've switched to handling `undef` instead of checking if the function is exported